### PR TITLE
preserve source positions for assertion rewriting

### LINKED ISSUE #12818: assertion rewriting ignores the original source positions
The assertion rewriting in pytest changes the source position for all nodes in the assertion to the source range of the assertion.

https://github.com/pytest-dev/pytest/blob/9515dfa58a144f3644fd29b256113d723c9c1955/src/_pytest/assertion/rewrite.py#L976-L978

This causes problems for tools like [executing](https://github.com/alexmojaki/executing) which uses these source ranges (since python 3.11) to provide a mapping back to the ast-node.

This causes the following problems.
- https://github.com/alexmojaki/executing/issues/2#issuecomment-553738141 executing does not work inside assertions.
- https://github.com/15r10nk/inline-snapshot/issues/57 I have to disable assertion rewriting to be able to find ast-nodes with executing.
- ... maybe more

I already created a pull-request to solve this problem #12815.

### PR DESCRIPTION
**Pull request by 15r10nk**
closes #12818

### CODE REVIEW

File: changelog/12818.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+assertion rewriting preserves the source ranges of the original instructions.
```
Comment (nicoddemus) [inline changelog/12818.bugfix.rst:1]:
```suggestion
Assertion rewriting now preserves the source ranges of the original instructions, making it play well with tools that deal with the ``AST``, like `executing <https://github.com/alexmojaki/executing>`__.
```

File: testing/test_assertrewrite.py [hunk 2]
```diff
@@ -131,10 +133,209 @@ def test_location_is_set(self) -> None:
...
-                assert n.col_offset == 0
-                assert n.end_lineno == 6
-                assert n.end_col_offset == 3
+                for location in [
+                    (n.lineno, n.col_offset),
+                    (n.end_lineno, n.end_col_offset),
+                ]:
+                    assert (3, 0) <= location <= (6, 3)
+
+    def test_positions_are_preserved(self) -> None:
```
Comment (nicoddemus) [inline testing/test_assertrewrite.py:142]:
```suggestion
    def test_positions_are_preserved(self) -> None:
        """Ensure AST positions are preserved during rewriting (#12818)."""
```

File: testing/test_assertrewrite.py [hunk 3]
```diff
@@ -131,10 +133,209 @@ def test_location_is_set(self) -> None:
...
+        def preserved(code):
```
Comment (nicoddemus) [inline testing/test_assertrewrite.py:143]:
```suggestion
        def preserved(code: str) -> None:
```

File: testing/test_assertrewrite.py [hunk 4]
```diff
@@ -131,10 +133,209 @@ def test_location_is_set(self) -> None:
...
+            s = textwrap.dedent(code)
+            locations = []
+
+            def loc(msg=None):
```
Comment (nicoddemus) [inline testing/test_assertrewrite.py:147]:
```suggestion
            def loc(msg: str|None=None) -> None:
```

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Thanks a lot @15r10nk for the PR, we really appreciate it!

Everything looks reasonable to me, appreciate the extensive tests.

Left a few minor suggestions, otherwise LGTM!

**Comment by 15r10nk** [comment]
thank you @nicoddemus for the review. I applied your requested changes.

File: src/_pytest/assertion/rewrite.py [hunk 5]
```diff
@@ -975,7 +975,8 @@ def visit_Assert(self, assert_: ast.Assert) -> list[ast.stmt]:
         # Fix locations (line numbers/column offsets).
         for stmt in self.statements:
             for node in traverse_node(stmt):
-                ast.copy_location(node, assert_)
+                if getattr(node, "lineno", None) is None:
```
Comment (nicoddemus) [inline src/_pytest/assertion/rewrite.py:978]: Can you leave a comment why this `if` is needed?

**Review by nicoddemus — APPROVED** [review]
LGTM, thanks!

**Comment by 15r10nk** [comment]
I commented it, squashed it and rebased it.

**Pull request merged**